llvm/include/llvm/CAS/ActionCache.h

```diff
@@ -0,0 +1,100 @@
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_CAS_CASACTIONCACHE_H
+#define LLVM_CAS_CASACTIONCACHE_H
+
+#include "llvm/ADT/Optional.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/CAS/CASReference.h"
+#include "llvm/Support/Error.h"
+
+namespace llvm::cas {
+
+class CASDB;
+class CASID;
+class ObjectProxy;
+
+/// A key for caching an operation.
+/// It is implemented as a bag of bytes and provided convenient constructor
+/// for string/CAS types.
+// TODO: Support CacheKey other than a CASID but rather any array of bytes.
+// To do that, ActionCache need to be able to rehash the key into the index,
+// which then `getOrCompute` method can be used to avoid multiple calls to
+// has function.
+class CacheKey {
+public:
+  StringRef getKey() const { return Key; }
+
+  CacheKey(const CASID &ID);
+  CacheKey(const ObjectProxy &Proxy);
+  CacheKey(const CASDB &CAS, const ObjectRef &Ref);
+
+private:
+  std::string Key;
+};
+
+/// A cache from a key describing an action to the result of doing it.
+///
+/// Actions are expected to be pure (write-once, collision is an error).
+class ActionCache {
+  virtual void anchor();
+
+public:
+  /// Get a previously computed result for \p ActionKey.
+  Optional<ObjectRef> get(const CacheKey &ActionKey) const {
+    return getImpl(arrayRefFromStringRef(ActionKey.getKey()));
+  }
+
+  /// Cache \p Result for the \p ActionKey computation.
+  Error put(const CacheKey &ActionKey, const ObjectRef &Result) {
+    return putImpl(arrayRefFromStringRef(ActionKey.getKey()), Result);
+  }
+
+  /// Cache \p Result using \p ObjectRef as a key.
+  Error put(const ObjectRef &RefKey, const ObjectRef &Result) {
+    CacheKey ActionKey(CAS, RefKey);
+    return putImpl(arrayRefFromStringRef(ActionKey.getKey()), Result);
+  }
+
+  /// Poison the cache for \p ActionKey.
+  Error poison(const CacheKey &ActionKey) {
+    return poisonImpl(arrayRefFromStringRef(ActionKey.getKey()));
+  }
+
+  /// Get or compute a result for \p ActionKey. Equivalent to calling \a get()
+  /// (followed by \a compute() and \a put() on failure).
+  Expected<ObjectRef> getOrCompute(const CacheKey &ActionKey,
+                                   function_ref<Expected<ObjectRef>()> Compute);
+
+  virtual ~ActionCache() = default;
+
+protected:
+  virtual Optional<ObjectRef> getImpl(ArrayRef<uint8_t> ResolvedKey) const = 0;
+  virtual Error putImpl(ArrayRef<uint8_t> ResolvedKey,
+                        const ObjectRef &Result) = 0;
+  virtual Error poisonImpl(ArrayRef<uint8_t> ResolvedKey) = 0;
+
+  ActionCache(CASDB &CAS) : CAS(CAS) {}
+
+  CASDB &getCAS() const { return CAS; }
+
+private:
+  CASDB &CAS;
+};
+
+/// Create an action cache in memory.
+std::unique_ptr<ActionCache> createInMemoryActionCache(CASDB &CAS);
+
+/// Create an action cache on disk.
+Expected<std::unique_ptr<ActionCache>> createOnDiskActionCache(CASDB &CAS,
```

Comment:
Is it supported to use the same Path for the action cache and the CAS? I would hate to have two separate command-line arguments for these paths.